ostd/libs/linux-bzimage/setup/src/x86/amd64_efi/efi.rs

```diff
@@ -55,9 +56,18 @@ fn efi_phase_boot(
 
     uefi_services::println!("[EFI stub] Relocations applied.");
 
-    uefi_services::println!("[EFI stub] Loading payload.");
     let payload = unsafe { crate::get_payload(&*boot_params_ptr) };
-    crate::loader::load_elf(payload);
+    const ELF_MAGIC_NUMBER: &[u8] = &[0x7F, 0x45, 0x4C, 0x46];
+    let kernel = match &payload[0..4] {
+        ELF_MAGIC_NUMBER => payload,
+        _ => {
+            uefi_services::println!("[EFI stub] Decompressing payload.");
+            &decompress_payload(payload)
```

Comment:
As mentioned above, `decompress_payload` can turn into `decode_payload`. Then the caller will not need to check the ELF magic here.